spinalcordtoolbox/scripts/sct_run_batch.py

```diff
@@ -175,6 +180,35 @@ def _find_nonsys32_bash_exe():
     return shutil.which('bash', path=nonsys32_paths)
 
 
+def _parse_dataset_directory(path_data, subject_prefix="sub-", ignore_ses=False):
+    """
+    Parse a dataset directory to find subject directories (or session subdirectories, if present).
+
+    Notes:
+        - The dataset is assumed to be structured in a (somewhat) BIDS-compliant way, see:
+          https://bids-specification.readthedocs.io/en/stable/02-common-principles.html#file-name-structure
+        - This function is a rudimentary version of the library PyBIDS: https://github.com/bids-standard/pybids.
+          TODO: https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3415
+    """
+    # Find subject directories within the root dataset folder
+    subject_dirs = sorted([f for f in os.listdir(path_data)
+                           if f.startswith(subject_prefix) and os.path.isdir(os.path.join(path_data, f))])
+
+    # Find session subdirectories within subject directories
+    session_dirs = []
+    for sub_dir in subject_dirs:
+        path_sub = os.path.join(path_data, sub_dir)
+        subject_session_dirs = sorted([os.path.join(sub_dir, f) for f in os.listdir(path_sub)
+                                       if f.startswith("ses-") and os.path.isdir(os.path.join(path_sub, f))])
+        session_dirs.extend(subject_session_dirs)
+
+    # Determine which directories to return
+    if session_dirs and not ignore_ses:
+        return session_dirs
+    else:
+        return subject_dirs
```

Comment:
Unfortunately, this is a significant behaviour change, and I believe it will affect existing datasets.

Specifically, if a dataset contains some `sub-` folders with no `ses-` subfolders, and other `sub-` folders with `ses-` subfolders, then:
* the old code returned a mix of `sub-` folders and `ses-` folders;
* the new code silently drops the `sub-` folders with no `ses-` subfolders.

I think you need something like a `result_dirs` variable which gets either an `append()` or an `extend()` for each `subject_dir`.

Reply:
Ahhhh, you're right! Thank you so much for catching this. :heart:

Reply:
I've reverted the rewrite from [2971227](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3862/commits/297122754a6cc5ad2809460dc0a73e3f85cbb425), then re-commited the smaller, more atomic changes that had been included in [2971227](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3862/commits/297122754a6cc5ad2809460dc0a73e3f85cbb425).